supernode/daemon/mgr/ha/etcd_tool.go

```diff
@@ -0,0 +1,164 @@
+package ha
+
+import (
+	"context"
+	"time"
+
+	"github.com/dragonflyoss/Dragonfly/supernode/config"
+
+	"github.com/sirupsen/logrus"
+	"go.etcd.io/etcd/clientv3"
+)
+
+// EtcdMgr is the struct to manager etcd.
+type EtcdMgr struct {
+	config            clientv3.Config
+	client            *clientv3.Client
+	leaseTTL          int64
+	leaseKeepAliveRsp <-chan *clientv3.LeaseKeepAliveResponse
+	hostIP            string
+	leaseResp         *clientv3.LeaseGrantResponse
+}
+
+const (
+	// ActiveSupernodeOFF means there is no active supernode.
+	ActiveSupernodeOFF = ""
+
+	// etcdTimeOut is the etcd client's timeout second.
+	etcdTimeOut = 10 * time.Second
+
+	// activeKeyPrefix is the keyPrefix of active supernode.
+	activeKeyPrefix = "/supernode/active/"
+	// standbyKeyPrefix is the keyPrefix of standby supernode.
+	//standbyKeyPrefix = "/supernode/standby/"
+)
+
+// the signal received from etcd watch.
+const (
+	// ActiveSupernodeChange means the active supernode is off.
+	ActiveSupernodeChange = 0
+	// ActiveSupernodeKeep means the active supernode is healthy.
+	ActiveSupernodeKeep = 1
+)
+
+// NewEtcdMgr produces a etcdmgr object.
+func NewEtcdMgr(cfg *config.Config) (*EtcdMgr, error) {
+	config := clientv3.Config{
+		Endpoints:   cfg.HAConfig,
+		DialTimeout: etcdTimeOut,
+	}
+	// build connection to etcd.
+	client, err := clientv3.New(config)
+	return &EtcdMgr{
+		hostIP: cfg.AdvertiseIP,
+		config: config,
+		client: client,
+	}, err
+}
+
+// WatchActiveChange is the progress to watch the etcd,if the value of key activeKeyPrefix changes,supernode will be notified.
+func (etcd *EtcdMgr) WatchActiveChange(ctx context.Context, messageChannel chan string) {
+	var watchStartRevision int64
+	watcher := clientv3.NewWatcher(etcd.client)
+	watchChan := watcher.Watch(ctx, activeKeyPrefix, clientv3.WithRev(watchStartRevision))
+	for watchResp := range watchChan {
+		for _, event := range watchResp.Events {
+			switch event.Type {
+			case ActiveSupernodeChange:
+				messageChannel <- string(event.Kv.Value)
+			case ActiveSupernodeKeep:
+				messageChannel <- ActiveSupernodeOFF
+			default:
+				logrus.Warnf("failed to watch active supernode,unexpected response: %d", int(event.Type))
+			}
+		}
+	}
+}
+
+// ObtainActiveInfo obtains the active supernode's information from etcd.
+func (etcd *EtcdMgr) ObtainActiveInfo(ctx context.Context, key string) (string, error) {
+	var value string
+	kv := clientv3.NewKV(etcd.client)
+	getRes, err := kv.Get(ctx, key)
+	if err != nil {
+		logrus.Errorf("failed to get the active supernode's(key: %s) info: %v", key, err)
+	}
+	for _, v := range getRes.Kvs {
+		value = string(v.Value)
+	}
+	return value, err
+}
+
+// ActiveResureItsStatus keeps look on the lease's renew response.
+func (etcd *EtcdMgr) ActiveResureItsStatus() {
+	for {
+		select {
+		case keepResp := <-etcd.leaseKeepAliveRsp:
```

Comment:
It's very weird that `etcd.leaseKeepAliveRsp` is assigned in method `etcd.KeepAlive`. The channel `leaseKeepAliveRsp` maybe changed to `nil` in another goroutine when current goroutine is waiting for receiving from it. A receive from nil channel blocks forever.